src/uproot/_dask.py

```diff
@@ -731,142 +743,302 @@ def _get_dask_array_delay_open(
+        return new_form, TrivialFormMappingInfo(new_form)
+
+
+T = TypeVar("T")
+
+
+class UprootReadMixin:
+    form_mapping: ImplementsFormMapping
+    base_form: Form
+    common_keys: set[str]
+    interp_options: dict[str, Any]
+
+    def read_tree(self, tree: HasBranches, start: int, stop: int) -> AwkArray:
+        assert start <= stop
+
+        from awkward._nplikes.numpy import Numpy
+
+        awkward = uproot.extras.awkward()
+        nplike = Numpy.instance()
 
-            return awkward.Array(
-                dask_awkward.lib.unproject_layout.unproject_layout(
-                    self.original_form,
-                    array.layout,
+        form, form_info = self.form_mapping(self.base_form)
+
+        # The remap implementation should correctly populate the generated
+        # buffer mapping in __call__, such that the high-level form can be
+        # used in `from_buffers`
+        mapping = form_info.load_buffers(
+            tree, self.common_keys, start, stop, self.interp_options
+        )
+
+        # Populate container with placeholders if keys aren't required
+        # Otherwise, read from disk
+        container = {}
+        for buffer_key, dtype in form.expected_from_buffers(
+            buffer_key=form_info.buffer_key
+        ).items():
+            # Which key(s) does this buffer require. This code permits the caller
+            # to require multiple keys to compute a single buffer.
+            keys_for_buffer = form_info.keys_for_buffer_keys({buffer_key})
+            # If reading this buffer loads a permitted key, read from the tree
+            # We might not have _all_ keys if e.g. buffer A requires one
+            # but not two of the keys required for buffer B
+            if all(k in self.common_keys for k in keys_for_buffer):
+                container[buffer_key] = mapping[buffer_key]
+            # Otherwise, introduce a placeholder
+            else:
+                container[buffer_key] = awkward.typetracer.PlaceholderArray(
+                    nplike=nplike,
+                    shape=(awkward.typetracer.unknown_length,),
+                    dtype=dtype,
                 )
-            )
 
-        return array
+        return awkward.from_buffers(
+            form,
+            stop - start,
+            container,
+            behavior=form_info.behavior,
+            buffer_key=form_info.buffer_key,
+        )
 
-    def project_columns(self, common_keys=None, original_form=None):
-        common_base_keys = self.common_base_keys
-        if self.form_mapping is not None:
-            awkward = uproot.extras.awkward()
+    @property
+    def meta(self) -> AwkArray:
+        awkward = uproot.extras.awkward()
+        high_level_form, form_info = self.form_mapping(self.base_form)
+        return awkward.typetracer.typetracer_from_form(
+            high_level_form,
+            highlevel=True,
+            behavior=form_info.behavior,
+        )
```

Comment:
The `meta` is separate to the column projection, but enables the `from_map` function to derive the meta from a single source of truth.